server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java

```diff
@@ -1206,6 +1207,52 @@ public void assertMatched() {
         }
     }
 
+    public void testReconfiguresToExcludeMasterIneligibleNodesInVotingConfig() {
+        final Cluster cluster = new Cluster(3);
+        cluster.runRandomly();
+        cluster.stabilise();
+
+        final ClusterNode chosenNode = cluster.getAnyNode();
+
+        assertThat(cluster.getAnyLeader().getLastAppliedClusterState().getLastCommittedConfiguration().getNodeIds(),
+            hasItem(chosenNode.getId()));
+        assertThat(cluster.getAnyLeader().getLastAppliedClusterState().getLastAcceptedConfiguration().getNodeIds(),
+            hasItem(chosenNode.getId()));
 
+        final boolean chosenNodeIsLeader = chosenNode == cluster.getAnyLeader();
+        final long termBeforeRestart = cluster.getAnyNode().coordinator.getCurrentTerm();
+
+        logger.info("--> restarting [{}] as a master-ineligible node", chosenNode);
+
+        chosenNode.close();
+        cluster.clusterNodes.replaceAll(cn -> cn == chosenNode ? cn.restartedNode(Function.identity(), Function.identity(),
+            Settings.builder().put(Node.NODE_MASTER_SETTING.getKey(), false).build()) : cn);
+        cluster.stabilise();
+
+        if (chosenNodeIsLeader == false) {
+            assertThat("term did not change", cluster.getAnyNode().coordinator.getCurrentTerm(), is(termBeforeRestart));
+        }
+
+        assertThat(cluster.getAnyLeader().getLastAppliedClusterState().getLastCommittedConfiguration().getNodeIds(),
+            not(hasItem(chosenNode.getId())));
+        assertThat(cluster.getAnyLeader().getLastAppliedClusterState().getLastAcceptedConfiguration().getNodeIds(),
+            not(hasItem(chosenNode.getId())));
+    }
+
+    public void testDoesNotPerformElectionWhenRestartingNonMasterNode() {
```

Comment:
would it make sense to generalize this test so that restarting any node that is not the current leader and without which there is still a publication quorum does not lead to an election?

We could then just call this `testDoesNotPerformUnnecessaryElection()`

Reply:
In fact I think the condition you described is a tautology: the term shouldn't change when restarting any node that isn't the current leader. If there's ≥3 master-eligible nodes then the cluster is fault tolerant; if there's 1 then we don't restart the unique master-eligible node, but if there's 2 then we know the leader is in the voting config and the other node isn't, so in all cases we can restart every node except the leader without triggering an election. Implemented this in 4f4cfe8d12a4d07760c931bd383fc94934b1e620.

Reply:
Hah, I thought about this, but dismissed the case of 2 nodes too quickly. You're right, cool that this simplifies the conditions.

Reply:
Bah except we randomly disable auto-shrinking the voting configuration so in a 2-node cluster we sometimes we have a 2-node config. Fixed that in a40f8f6ff8f551efeb65c95870560cd01a44ef29.

Reply:
💃